### PR TITLE
process: rewrite process management

### PR DESCRIPTION
**Pull request by rogpeppe**
We rewrite the process management to use a more
disciplined approach to concurrency. There's a main
goroutine for each process that monitors any process
that's currently active, and another goroutine that's
responsible for coordinating updates to the configuration.

Logging isn't hooked up and restarts aren't tested at all,
but this is probably a good stopping point for a review.

### CODE REVIEW

File: supervisor.go [hunk 1]
```diff
@@ -41,9 +42,37 @@ func NewSupervisor(configFile string) *Supervisor {
...
 
+// Close closes the server, stopping all processes gracefully.
+func (s *Supervisor) Close() error {
+	// Updating with the empty configuration should gracefully stop all processes
+	// in the correct order and stop and tear down any current network
+	// requests.
+	if err := s.procMgr.Update(context.TODO(), &config.Config{}); err != nil {
+		zap.L().Info("cannot terminate processes", zap.Error(err))
+	}
+	s.mu.Lock()
```
Comment (rogpeppe) [inline supervisor.go:58]: defer unlock

Comment (stuartcarnie) [reply supervisor.go:58]: Where is this deferred?

**Review by stuartcarnie — COMMENTED** [review]
Looks good to me, @rogpeppe – you had a comment, "defer unlock", but it isn't clear to me what that is for. I didn't see an unlock for the mutex in that function, so is it missing or is it coordinated elsewhere?

File: supervisor.go (hunk 1 above)
Comment (rogpeppe) [reply supervisor.go:58]: it's not! it was me in a concert with minimal connectivity perusing the code and commenting to remind myself that we must unlock the mutex here.

Comment (rogpeppe) [reply supervisor.go:58]: Fixed

**Review by stuartcarnie — APPROVED** [review]
This is a huge PR, with lots of great improvements. I haven't reviewed it in considerable depth, as I consider this largely a rewrite. You've created a TODO list and adding lots of tests which is good enough for me.

**Pull request merged**